src/util/config.js

```diff
@@ -1,83 +1,146 @@
+		},
+		port: {
+			format: 'port',
+			default: 8000
+		}
+	},
+	oauth: {
+		auth_url: {
+			format: 'url',
+			default: 'https://secure.meetup.com/oauth2/authorize'
 		},
-	};
-	config.duotoneUrls = getDuotoneUrls(duotones, config.PHOTO_SCALER_SALT);
-	config.API_SERVER_ROOT_URL = `${config.API_PROTOCOL}://${config.API_HOST}`;
+		access_url: {
+			format: 'url',
+			default: 'https://secure.meetup.com/oauth2/access'
+		},
+		secret: {
+			format: function (secret) {
+				if (secret.toString().length < 1) {
+					throw oauthError;
+				}
+			},
+			default: ''
+		},
+		key: {
+			format: function (key) {
+				if (key.toString().length < 1) {
+					throw oauthError;
+				}
+			},
+			default: ''
+		}
+	},
+	photo_scaler_salt: {
+		format: function (salt) {
+			if (salt.toString().length < 1) {
+				throw new Error('get PHOTO_SCALER_SALT from #web-platform team');
+			}
+		},
+		default: ''
+	}
+});
+
 
-	// currently all config is available syncronously, so resolve immediately
-	return Promise.resolve({ ...config, ...overrideConfig })
-		.then(validateConfig);
+// Optionally override these properties with a JSON file
+const env = config.get('env');
+const configFile = `./config.${env}.json`;
+if (fs.existsSync(configFile)) {
+	config.loadFile(configFile);
 }
 
-function validateConfig(config) {
-	const oauthError = new Error('get oauth secrets from web platform team');
-	const configSchema = Joi.object().keys({
-		API_PROTOCOL: Joi.any().only(['https', 'http']).required(),
-		API_HOST: Joi.string().hostname().required(),
-		API_SERVER_ROOT_URL: Joi.string().uri(),
-		API_TIMEOUT: Joi.number(),
-		COOKIE_ENCRYPT_SECRET: Joi.string().min(32).required().error(
-			new Error('set COOKIE_ENCRYPT_SECRET env variable to a random 32+ character string')
-		),
-		CSRF_SECRET: Joi.string().min(32).required().error(
-			new Error('set CSRF_SECRET env variable to a random 32+ character string')
-		),
-		DEV_SERVER_PORT: Joi.number().integer().max(65535),
-		OAUTH_AUTH_URL: Joi.string().uri().required(),
-		OAUTH_ACCESS_URL: Joi.string().uri().required(),
-		PHOTO_SCALER_SALT: Joi.string().min(1).required().error(
-			new Error('get PHOTO_SCALER_SALT from web platform team')
-		),
-		oauth: Joi.object().keys({
-			secret: Joi.string().min(1).required().error(oauthError),
-			key: Joi.string().min(1).required().error(oauthError),
-		}).required(),
-		duotoneUrls: Joi.object(),
-	}).required();
+// Load environment dependent configuration
 
-	const result = Joi.validate(config, configSchema);
-	if (result.error) {
-		throw result.error;
-	}
-	return result.value;
-}
+config.set(
+	'duotone_urls',
+	getDuotoneUrls(
+		duotones,
+		config.get('photo_scaler_salt')
+	)
+);
+
+config.set(
+	'api_server_root_url',
+	`${config.get('api.protocol')}://${config.get('api.host')}`
+);
+
+config.set(
+	'isProd',
+	config.get('env') === 'production'
+);
+
+config.set(
+	'isDev',
+	config.get('env') === 'development'
+);
+
+config.validate();
 
+export default config;
```

Comment:
it's fine to roll back the Promise setup - it was a vestige of a time where the server had to make some HTTP requests to get some config data when booting, so the config object had to be built async. I think we can avoid that requirement in the future, but I had originally kept the `Promise` just in case we needed it.

Treating the config object as a Plain Old Javascript Object throughout the rest of the code would definitely make it easier to understand - the getters and setters feel a little awkward to me, although I'm sure 90% of that is just not being used to node convict's API. I can't think of a good reason not to export the POJO version, though.